Digital Credentials: implement CANONICAL_REQUEST_OBJECTS with actual mDoc data

marcoscaceres opened the pull request:
Part 4:

- Add CANONICAL_REQUEST_OBJECTS with real mDoc deviceRequest and encryptionInfo data
- Update factory functions to use canonical objects as defaults
- Refactor internal architecture with new helpers
- Make mediation optional in type definitions (no longer defaults to 'required')
- Add comprehensive data parameter support to config interfaces
- Complete property bag + signal + protocol selection + canonical objects integration

marcoscaceres committed:
Digital Credentials: implement CANONICAL_REQUEST_OBJECTS with actual mDoc data

- Add CANONICAL_REQUEST_OBJECTS with real mDoc deviceRequest and encryptionInfo data
- Update factory functions to use canonical objects as defaults with data spread
- Refactor internal architecture with new buildRequests and makeOptionsInternal helpers
- Make mediation optional in type definitions (no longer defaults to 'required')
- Add comprehensive data parameter support to config interfaces
- Complete property bag + signal + protocol selection + canonical objects integration

marcoscaceres commented on digital-credentials/support/helper.js:
```diff
@@ -10,113 +10,229 @@
+      data = { ...CANONICAL_REQUEST_OBJECTS["openid4vp-v1-multisigned"] },
+    ) => {
+      return { protocol: "openid4vp-v1-multisigned", data };
+    },
```
@copilot let's got with option 2. Make a suggestion here.

mohamedamir commented on digital-credentials/support/helper.js:
```diff
@@ -34,100 +45,162 @@ const SUPPORTED_CREATE_PROTOCOL = CREATE_PROTOCOLS.find(
+ * @returns {Req[]}
+ * @throws {Error} If an unknown protocol string is encountered.
+ */
+function buildRequests(protocols, mapping, data) {
```
Maybe we should define a new `Request` that contains both `Protocol` and `data`?

marcoscaceres commented on digital-credentials/support/helper.js:
```diff
@@ -34,100 +45,162 @@ const SUPPORTED_CREATE_PROTOCOL = CREATE_PROTOCOLS.find(
   },
   create: {
-    "openid4vci": () => makeOID4VCIDict(),
+    openid4vci: (data = { ...CANONICAL_REQUEST_OBJECTS["openid4vci"] }) => {
```
I think Prettier removes the quotes.

marcoscaceres committed:
Digital Credentials: Simplify helper function architecture and add requests parameter

- Add requests parameter to MakeGetOptionsConfig and MakeCreateOptionsConfig interfaces
- Replace complex makeOptionsUnified/makeOptionsInternal system with simpler makeCredentialOptionsFromConfig
- Convert buildRequestsFromProtocols to functional programming approach using .map()
- Remove factory functions (makeOID4VPDict, makeMDocRequest) in favor of inline functions in allMappings
- Make mediation optional in CredentialRequestOptions and CredentialCreationOptions
- Enhance TypeScript annotations with generics for better type safety
- Support empty protocol arrays (protocol: []) for testing scenarios
- Clean up unused imports and improve documentation alignment

marcoscaceres commented:
In the future, we could combine them:

```JS
makeGetRequest({requests: [orangeRequest, "openid4vp-v1-unsigned", bananaRequest, "org-iso-mdoc"] });
```

That can be a nice followup.

marcoscaceres committed:
Add data parameter to config interfaces and update tests to use clean data override pattern

- Add data parameter to MakeGetOptionsConfig and MakeCreateOptionsConfig for overriding canonical data
- Wire explicitData parameter through buildRequestsFromProtocols call chain properly
- Update tests to use clean makeGetOptions({data}) pattern instead of manual object construction
- Replace post-construction data manipulation with clean data override during construction
- Maintain separation of concerns: protocol selection automatic, data override optional

mohamedamir commented on digital-credentials/support/helper.js:
```diff
@@ -27,107 +26,145 @@ const GET_PROTOCOLS = /** @type {const} */ ([
-    throw new Error(`Internal error: Invalid options type specified: ${type}`);
-  }
+function makeCredentialOptionsFromConfig(config, mapping) {
+  const { protocol, requests = [], data, mediation = "required", signal } = config;
```
IMHO, we should also remove populating the default mediation to make sure browsers don't reply on it being set.

mohamedamir commented on digital-credentials/dc-types.ts:
```diff
@@ -33,9 +33,20 @@ export interface MobileDocumentRequest {
+  /**
+   * Optional data to override canonical data for protocol-based requests.
+   */
+  data?: MobileDocumentRequest | object;
```
I recall that some tests are testing passing here invalid data, which aren't necessarily objects.
Maybe we should make this to capture that?

```suggestion
  data?: MobileDocumentRequest | any;
```

mohamedamir reviewed (approved):
This looks solid now!
Thanks a lot Marcos for the revisions!
Some minor comments are left!
But otherwise, LGTM
